### DISCUSSION TITLE
Multi project and answers inheritance

### DISCUSSION THREAD

**Discussion by dcarrillo** [copier-org/copier#416] [Q&A] [answered]
Hello, at https://www.adevinta.com we're evaluating `copier` to replace several projects based on `Cookiecutter`.

We want to share templates among projects having a base template, so we want to ask the questions once and the answers will be shared by the other templates.

```text
base template (Questions: project_name, team, tenant, etc)
     -> pipeline.yaml.jinja
     -> .travis.yaml.jinja

kotlin microservice template <- Answers from base template
     -> src/java.x.x.x/main.kt.jinja

python microservice template <- Answers from base template
     -> module/__init__.py.jinja
```

We've read a similar question at https://github.com/copier-org/copier/discussions/402 but according to the doc https://copier.readthedocs.io/en/latest/configuring/#applying-multiple-templates-to-the-same-subproject in order to use multi-template, each template should have its own answers file.

We were able to do the trick by "sharing" the answers file generated by the base template, but not sure if this is the right approach.

Is it possible this kind of multi-project ? 

Thanks in advance.

**Comment by yajo** [comment, marked as answer]
> Hello, at https://www.adevinta.com we're evaluating `copier` to replace several projects based on `Cookiecutter`.

Glad to hear that, I hope you like it 😊 

> We've read a similar question at #402 but according to the doc https://copier.readthedocs.io/en/latest/configuring/#applying-multiple-templates-to-the-same-subproject in order to use multi-template, each template should have its own answers file.

Yes, that's right.

One thing is applying multiple templates to the same project, and a different topic is template inheritance.

Right now, Copier only supports obtaining answers from a single file, or from arguments. Also, the answers file is attached to a template, to be able to support autoupdates.

So, the 1st thing you should ask yourself is if your templates are related, or if they can be detached. If they can be detached, then put each specific question in the proper template, and evolve them separately.

Next, ask yourself if your templates will support updates. If they are bootstrap-only (like Cookiecutter), then you don't really need to have an answers file, because the template will never actually reuse those answers.

If you find out that none of these approaches solves your problem, then another resource you can use is [to include other yaml files](https://copier.readthedocs.io/en/latest/configuring/#include-other-yaml-files) in the main `copier.yaml` file of your template. This way, you can split question sections in different files, put them in a git submodule that can be shared between templates, and then include the questions you need on each template. The downside would be that you'd have to repeat the answers the 1st time you bootstrap each template.

Another option is to use tasks to always propagate answers from the main template to the other used templates. That's custom code, of course.

Finally, if none of these work for you, then you're asking for a feature Copier doesn't have right now. If you feel like volunteering a PR, we can discuss the design.

> We were able to do the trick by "sharing" the answers file generated by the base template, but not sure if this is the right approach.

I'm not sure that will work reliably. Possibly, when updating, the new template will override the answers file. In any case, it's not a supported use case.

**Reply by dcarrillo** [reply]
Thank you @Yajo for the explanation.

We've evolved a little bit from our first approach and now we are using git submodules joined in a single template, so there is only a set of copier questions (and answers). This works for us.

Kind regards.

PS We're looking forward to 6.0 GA release :)
